oio/crawler/integrity.py

```diff
@@ -0,0 +1,334 @@
+
+        self.list_cache[account] = containers
+        self.running[account].send(True)
+        del self.running[account]
+        self.accounts_checked += 1
+
+        if recurse:
+            for container in containers:
+                t = target.copy()
+                t.container = container
+                self.pool.spawn_n(self.check_container, t, True)
+
+        if error and self.error_file:
+            self.write_error(target)
+        return containers
+
+    def check(self, target):
+        if target.chunk and target.obj and target.container:
+            self.pool.spawn_n(self.check_chunk, target)
+        elif target.obj and target.container:
+            self.pool.spawn_n(self.check_obj, target, True)
+        elif target.container:
+            self.pool.spawn_n(self.check_container, target, True)
+        else:
+            self.pool.spawn_n(self.check_account, target, True)
+
+    def wait(self):
+        self.pool.waitall()
+
+    def report(self):
+        def _report_stat(name, stat):
+            print("{0:18}: {1}".format(name, stat))
+
+        print()
+        print('Report')
+        _report_stat("Accounts checked", self.accounts_checked)
+        if self.account_not_found:
+            _report_stat("Missing accounts", self.account_not_found)
+        if self.account_exceptions:
+            _report_stat("Exceptions", self.account_not_found)
+        print()
+        _report_stat("Containers checked", self.containers_checked)
+        if self.container_not_found:
+            _report_stat("Missing containers", self.container_not_found)
+        if self.container_exceptions:
+            _report_stat("Exceptions", self.container_exceptions)
+        print()
+        _report_stat("Objects checked", self.objects_checked)
+        if self.object_not_found:
+            _report_stat("Missing objects", self.object_not_found)
+        if self.object_exceptions:
+            _report_stat("Exceptions", self.object_exceptions)
+        print()
+        _report_stat("Chunks checked", self.chunks_checked)
+        if self.chunk_not_found:
+            _report_stat("Missing chunks", self.chunk_not_found)
+        if self.chunk_exceptions:
+            _report_stat("Exceptions", self.chunk_exceptions)
+
+
+def main():
+    parser = argparse.ArgumentParser(description="Check integrity")
+    parser.add_argument('namespace', help='namespace name')
+    parser.add_argument('target', metavar='T', nargs='*',
+                        help='target to check integrity')
+    parser.add_argument('-o', '--output', help='output file')
+    parser.add_argument('-v', '--verbose',
+                      action='store_true', help='verbose output')
+
+    args = parser.parse_args()
+
+    checker = Checker(args.namespace, error_file=args.output)
+    if not os.isatty(sys.stdin.fileno()):
+        source = sys.stdin
+    else:
+        source = cStringIO.StringIO(' '.join(args.target))
+    args = csv.reader(source, delimiter=' ')
```

Comment:
I suggest to make this delimiter configurable. A single space will cause problems with contents (or containers) with a space in their name.